// minecraft-server-util/dist/structure/Description.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ansi_styles_1 = __importDefault(require("ansi-styles"));
const formattingCode = /\u00C2?\u00A7([a-fklmnor0-9])/g;
const ansiMap = new Map();
ansiMap.set('0', ansi_styles_1.default.black);
ansiMap.set('1', ansi_styles_1.default.blue);
ansiMap.set('2', ansi_styles_1.default.green);
ansiMap.set('3', ansi_styles_1.default.cyan);
ansiMap.set('4', ansi_styles_1.default.red);
ansiMap.set('5', ansi_styles_1.default.magenta);
ansiMap.set('6', ansi_styles_1.default.yellow);
ansiMap.set('7', ansi_styles_1.default.gray);
ansiMap.set('8', ansi_styles_1.default.blackBright);
ansiMap.set('9', ansi_styles_1.default.blueBright);
ansiMap.set('a', ansi_styles_1.default.greenBright);
ansiMap.set('b', ansi_styles_1.default.cyanBright);
ansiMap.set('c', ansi_styles_1.default.redBright);
ansiMap.set('d', ansi_styles_1.default.magentaBright);
ansiMap.set('e', ansi_styles_1.default.yellowBright);
ansiMap.set('f', ansi_styles_1.default.whiteBright);
ansiMap.set('k', ansi_styles_1.default.reset);
ansiMap.set('l', ansi_styles_1.default.bold);
ansiMap.set('m', ansi_styles_1.default.strikethrough);
ansiMap.set('n', ansi_styles_1.default.underline);
ansiMap.set('o', ansi_styles_1.default.italic);
ansiMap.set('r', ansi_styles_1.default.reset);
/**
 * The result of the formatted description of the server.
 * @class
 */
class Description {
    /**
     * Creates a new description class from the text.
     * @param {string} descriptionText The MOTD text
     * @constructor
     */
    constructor(descriptionText) {
        this.descriptionText = descriptionText;
    }
    /**
     * Converts the MOTD into a string format
     * @returns {string} The string format of the MOTD
     */
    toString() {
        return this.descriptionText;
    }
    /**
     * Converts the MOTD into a string format without any formatting
     * @returns {string} The MOTD string without formatting
     */
    toRaw() {
        return this.descriptionText.replace(formattingCode, '');
    }
    /**
     * Converts the special formatting characters to ANSI escape codes, commonly used for terminal formatting
     * @returns {string} The ANSI escaped formatting text
     */
    toANSI() {
        return this.descriptionText.replace(formattingCode, (match, p1) => {
            const value = ansiMap.get(p1);
            if (!value)
                return ansi_styles_1.default.reset.open;
            return value.open;
        }) + ansi_styles_1.default.reset.open;
    }
}
exports.default = Description;
